### ISSUE TITLE
[Proposal] Resolve Services on "this" Constructor or Static Constructor

### ISSUE THREAD

**Issue by RandyBuchholz** [aspnet/DependencyInjection#554]
## Internal DI System Scope
DI is something of an all or nothing deal - if your object wasn't created by the container you have no access to the DI services. Once you "new" something or use things like Activator.Create() you are out of the DI scope. So is everything downstream, limiting the use of DI more towards upstream and business service classes, when it can be just as valuable to all classes.

As a result it creates cases where you may need to use static or instance classes when you should be able to use DI. A common case is utilities. Many utilities are implemented as static classes and methods. This makes them tightly coupled and difficult to change and test. Rewriting them as utility services and injecting them would be great, but unless the classes that use them are part of the DI hierarchy you can't resolve them.
 
This proposal provides a way to use the internal DI system with any class. This extends DI beyond top-level business service classes all the way down to small runtime objects.

### Example
Say I have a set of features defined by `IFooProcesser` that accept `Foo` as a parameter, and manipulate Foo's.
```
public interface IFooProcessor{
    Foo Reverse(Foo inFoo);
    void Transform(ref Foo foo);
}
```
IFooProcessor provides "utility services" and not business services, so they may be needed in many places, both business and "data". I also want to be able to swap them out. Foo's are runtime objects, and the type `Bar` can contain them. 
```
public class Bar {
   public List<Foo> foos { get; set; }
}
```
I can use my IFooProcessor on my Foo's.
````
public class Bar{
    private readonly IFooProcessor _fooProc;
    
    public Bar(IFooProcessor fp){
        _fooProc = fp;
    }
    
    public List<Foo> foos { get; set; }

    public Foo Oof(Foo foo) => _fooProc.Reverse(foo);
    
    public void ChangeThem(){
        foreach(var f in foos){
            _fooProc.Transform(ref f);
        }
}
````
The problem is `Bar` is also a runtime object, and `Bar` participates in things like serialization and Activator.Create() that require a public parameterless constructor. Built-in DI only allows a single constructor. I need to be able to do `var f = new Bar()`. Once I do though, I loose my Services. I have to implement `public Bar() { }` so it becomes
```
public class Bar{
   private readonly IFooProcessor _fooProc;
   public Bar() {
    ...
   }
}
```
and now my service IFooProcessor  can't resolve.

IFooProcessor is a basic utility-type service. These are often implemented as static classes so they are easy to use anywhere.  Going that route though doesn't allow implementing the interface (`public static class X : IX` is not valid). If I go with an instance class I'm `new`ing or resorting to "creator" approaches, when I should just register it as a DI singleton and be done. Many IFooProc type of classes are either "in" or "out" of DI, - they can't easily participate in both. Having a class be able to participate in DI and be "newed"  would be valuable.

The ability to have multiple "DI Execution Roots" would also be valuable. These would share the same Composition Root, (Though multiple composition trees could be supported.) but can be placed anywhere in the system. Essentially, a class I new x(), Activator.Create(), or similar can be tagged as being another "root' in a DI tree/chain. I can see two ways to do this.

### The "this" Constructor
This approach modifies the way "this" works on a constructor to resolve services from there. The syntax could be like the current syntax and call another constructor. This is like "Poor Mans Injection". DI would just inspect the "this" part of the constructor as part of its process and the parameterless constructor is there for other needs.
```
public class Bar{
    private readonly IFooProcessor _fooProc;

    public Bar() : this( IFooProcessor fooProcessor)

    private Bar(IFooProcessor fooProcessor){
        _fooProc = fooProcessor;
   }
}
```
This provides both the ability to "new" and inject at the same time, making it easy to create and inject utility services. This has also created a new "Service Root" on `Bar`. If the FooProcessor implementation has dependencies they will inject just like normal. This will work fine for 
```
public class FooProcessor : IFooProcessor {
    private readonly IServiceDependency _sdp;

    public FooProcessor(IServiceDependency depend) {
        _sdp = depend;
    }
}
```
The new Service Root on Bar extends down to FooProcessor and beyond. 

A more explicit syntax might be better. This would be nice candy even for regular DI.
```
public class Bar{
    private readonly IFooProcessor _fooProc;

    public Bar() : this ( IFooProcessor => _fooProc )
        or
    public Bar() : this ( _fooProc <= IFooProcessor )
}
```
### Static Constructor
This is a deep change that effects language fundamentals. It is more of a thought/logical approach than an actual suggestion. With this change, static constructors would be able to accept parameters and set readonly fields. Parameters would be limited to DI-resolvable items. 

An implementation approach could just defer the actions to the instance constructor and not actually change the underlying implementation of the static constructor.

Considering two things about constructor injection DI it's logical for DI to be implemented at this (creation) level. Putting the injected items in the static constructor makes clear what the class needs to function. If we accept/assume that most injected classes are essential to the class (part of the class composition) they may belong in the static constructor, at least conceptually. It also provides a level of structural vs business/functional separation. Core composition structure is static, functional/business is instance.

Using the shorthand syntax it could look like this.
```
public class Bar{
    private readonly _fooProc;
    
    // "Structural" construction
    static Bar( IFooProcessor => _fooProc) { }
    
   // Cleaner "functional" constructors
   public Bar() { }
   public Bar(sting name){ }
}
```
Again, the static constructor approach may more conceptual sense than language sense.

**Comment by davidfowl** [comment]
> Built-in DI only allows a single constructor

It doesn't, not sure where you got that premise from. The built in DI container prefers the constructor with the most resolvable services. See https://github.com/aspnet/DependencyInjection/blob/rel/1.1.1/src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs#L640 for details.

It seems like what you are suggesting requires new language syntax and potentially runtime features. Was this more of an "out there" brainstorming type suggestion? (just trying to understand so I can formulate the appropriate reply).

**Comment by RandyBuchholz** [comment]
Hmmm, don't know where I got the single constructor from it's been in my head forever though. Could that have been the case in the beta days? 

This is a real world situation I've run into a couple of times and can't find a solution. But I may have missed a few approaches with the single constructor thing stuck in my head...

I'll put a repo together to show the problem. In the simplest form, a key issue is trying to dynamically create a DI enabled class using Activator.Create().

I have a class that uses several static utility classes. These would be better as injectables.

**Comment by RandyBuchholz** [comment]
Hmmm, don't know where that came from, I've thought that forever. Was it ever the case in the Beta days? Feel dumb...

No this is a not just an out of the blue thing, but something I've struggled with many times. It's sort of a bottom-up problem. So at the heart, say I have low-level utility classes that are static. I have a class that uses the utility classes - `Project`. The utilities should be "replaceable" so I abstract them to interfaces, and implement instance class. Now I can program the consuming class using the interfaces and manually add `private IUtilty _util = new Utility();` to `Project`. In my "parent" class `Portfolio` I do `var p = new Project()` and everything works. So far, so good. But I have newed-up `Utiltiy` in Project and every other consumer class. Maybe DI is the way to go. I register the interface and instance, modify Project
```
public class Project {
    private IUtility _util;   
    public Project ( IUtility util ){
        _util = util;
    }
}
```
Of course now `var p = new Project()` fails - it wants a parameter. `var p = new Project(new Utility());` Why? Because I'm not connected to the ServiceProvider. This is because `Portfolio` is not connected to a provider. It seems ServiceProvider isn't pervasive across my application, where it will magically resolve any constructor, but is part of a hierarchy. It will only resolve if the parent was resolved. So I convert my classes all the way up the chain until I reach one in the hierarchy. Now everything works and resolves. As long as I stay in the chain - don't use new (), that breaks the chain. This is my understanding and experience with DI so far.

The problems start when you have to "new" something. In my case I have to have an instance of a generic type created when I create a containing type. This is for model binding. I create the instance using Activator.Create(). This is where things fall apart. Activator.Create() will take parameters, but it doesn't (seem to) work with DI. I get an instance, but none of the injections resolve.

The idea was making DI more pervasive by having an approach that will let DI recognize it is wanted from anywhere. In the case I seem to encounter I want DI for business services (top down) and I want DI for shared utilities (bottom-up), and still be able to new/actvator/deserialize my operating classes (that utilize injected utilities).

Hope that made sense, or maybe I'm missing something basic here.

**Comment by davidfowl** [comment]
> Of course now var p = new Project() fails - it wants a parameter. var p = new Project(new Utility()); Why? Because I'm not connected to the ServiceProvider. This is because Portfolio is not connected to a provider.

It doesn't need to fail, you can have multiple ctors but you still won't be able to use the DI container if you don't create instances from the service provider. There's no getting around that.

> It seems ServiceProvider isn't pervasive across my application, where it will magically resolve any constructor, but is part of a hierarchy

There's no magic to it. The only reason it seems magical is because you're using a framework that's doing the heavy lifting on your behalf. If you're in a place where the framework doesn't control activation (which is the case for lots of utility code then you're the one that needs to call GetService). 

> The problems start when you have to "new" something. In my case I have to have an instance of a generic type created when I create a containing type. This is for model binding. I create the instance using Activator.Create(). This is where things fall apart. Activator.Create() will take parameters, but it doesn't (seem to) work with DI. I get an instance, but none of the injections resolve.

Right, the DI container is supposed to "new" it up for you, it's basically all or nothing (as you've eluded to). 

> The idea was making DI more pervasive by having an approach that will let DI recognize it is wanted from anywhere. In the case I seem to encounter I want DI for business services (top down) and I want DI for shared utilities (bottom-up), and still be able to new/actvator/deserialize my operating classes (that utilize injected utilities).

I understand your request but I just don't understand concretely how it would work. It's more of an interesting idea but I don't think it's feasible without more radical changes to other parts of the stack. What you're asking for would basically be possible if the runtime (or compiler) had a hook that would give up control of any object creation and maybe the DI system could hook into it. Might be something to propose on http://github.com/dotnet/coreclr. Even with that hook the DI container would need to be ephemerally available so that the generate code could call into it from a utility class.

**Comment by RandyBuchholz** [comment]
David, thanks for you thorough response.

Yeah, to do it right there would need to be some sort of "creating" event on all objects that could be intercepted - a major change that would add overhead to just about everything that happens. A half-step could be a pre-compile parse that converts the assignment syntax into manual `GetService()` calls. More of a hack though. btw - forgot to tag "magically" as sarcasm.

ActivatorUtilities is getting me down the path. Thanks for that.

Do you have a recommended approach to getting the default service provider in non-DI classes?

Right now I'm creating a "Global ServiceProvider Factory" as a static class in the application/assembly root namespace
```
namespace Projects {
    public static class GlobalServices {
        public static IServiceCollection Services { get; set; } 
        public static IServiceProvider Provider { get => Services.BuildServiceProvider(); }
    }
}
```
I set the collection at the end of  `Startup.ConfigureServices` with `GlobalServices.Services = services;` Then I just use `GlobalService.Provider.GetService<ServiceType>()`.

Is that an OK approach or is there something better?

### ______________________________________________

Here is a little out-there thinking - ServiceBrokers
I could extend GlobalServices to handle scoped service sets. Like having `UtilityProvider` that only registers the services needed within that scope. When I create things like libraries, domains, or major namespaces I can expose a services collection at the root of them. These could even register as part of the package. ServiceSets can implement "interfaces of services" or `IServiceProvider`'s. When I manually "inject" a service into a class, all of the classes below it will be using that version of the service set. 

This could be used in things like Domain Driven concepts. I may be creating bounded contexts that use the same features, but in a different context. I can create an interface for the service combinations required. Within a namespace I create a service provider that implements that interface.
```
public interface IDomainServices {
   IScopedDataReader DataReader { get; set; }
   IScopedDataWriter DataWriter { get; set; }
   IScopedPeople People { get; set; }
}

public interface IScopedPeople {
    IEnumerable<People> Owners { get; set; }
}

// Conceptual - Probably confusing
// 
public interface IBenefitsDomainServices : IDomainServices { 
   public interface IScopedPeople People : IBenefitPeople : IScopedPeople {
       ...
   }
}

public interface IRisksDomainServices : IDomainServices { 
   public interface IScopedPeople People :   IRiskPeople : IScopedPeople {
       ...
   }
}
```

I can bind these using Attributes.
A lot is missing here, but that leads to things like:
```
public class People {
   private readonly IScopedDataReader _reader;
   public People (IScopedDataReader reader)
   ...

   public IEnumerable<People> Owners => ...
}

[ServiceScope("BenefitsDomain")
public class Benefit { 
    private IScopedPeople _people;
    public Benefit() { }
    ...
    public People => _people.Owners; 
}

[ServiceScope("RisksDomain")
public class Risk {
    private IScopedPeople _people;
    public Risk() { }
    ...
    public People => _people.Owners;   
}
```
In this case, `Benefit` and `Risk` don't participate in upper-level DI. And are just newed. The attribute causes the constructor to "inject" a Service Composition Root that implements specific versions of services. The classes below them will now get the scoped versions of these interfaces/services. Risk is using "RiskPeopleServices" and Benefit "BenefitPeopleServices" for _people. The application has multiple Composition Roots, and each tree can inject different implementations of the same interface. In this case `Owners` comes from `People` in both `Risk` and `Benefit`, but they are resolved differently. Based on the `IService`'s specified in `IRisksDomainServices` and `IBenefitDomainServices`. Like I said, out-there...

**Comment by RandyBuchholz** [comment]
@davidfowl I've looked at constructor overloading and DI and know where I got the single constructor perception from. As you said, DI resolves to the most resolvable constructor. The link to the testing implies it should look for the "least" resolvable - i.e., longest resolvable signature. It is currently resolving to the "most resolvable", so if you have a parameterless constructor it will resolve to that.

So while the class can have multiple constructors, from a DI perspective there is only one - the one with the minimum signature.

I have a repo here to show the problem.
https://github.com/Randy-Buchholz/DIOverload/tree/master/DIOverload

I have opened an issue on this
Resolves to Least Complex Signature, not Most Complex #557
